transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -723,7 +747,17 @@ protected void doClose() throws Exception {
 
     @Override
     protected void doBeginRead() throws Exception {
-        // NOOP
+        // read from tempInboundMessages and fire channelRead.
+        if (!tempInboundMessages().isEmpty()) {
+            Object msg;
+            while ((msg = tempInboundMessages().poll()) != null) {
+                ChannelPipeline p = pipeline();
+                p.fireChannelRead(msg);
+            }
+
+            tempInboundMessages().clear();
```

Comment:
why you call `clear()` here ? This should not be needed as you already polled from it.

Reply:
fixed.